### PR TITLE
feat: new (pfe-readtime) to display the estimated time to read

### PR DESCRIPTION
**Pull request by alazzara**
## <Pull request description>
New web component for calculating the read time of a section of a page. This can be done by getting a value inside a data attribute or by capturing the number of words in a given section and dividing it by the read rate.

New web component for calculating the read time of a section of a page.

### Preview

Link(s) to demo page(s) where this element can be viewed:
- [Link](https://deploy-preview-1510--happy-galileo-ea79c4.netlify.app/elements/pfe-readtime/demo/) 

### Testing instructions

1. If you add a `word-count` data attribute to `pfe-readtime` it should use that to calculate the readtime.
2. Example: word-count="250" should result in a readtime of 1-minute
3. If you do not add a `word-count` attribute to `pfe-readtime` it should count the words in the section and do the calculation based on that.
4. If the readtime is less than 1 minute the pfe-readtime should not render on the page.
5. If you change the language code to zh it should change the wpm readtime to 158.
6. If you change word-count="158" on english it should not show the readtime. When you change the language code to `zh` it should say 1-minute readtime.

#### Browser requirements

Your component should work in all of the following environments:

- [x] Latest 2 versions of Edge
- [x] Internet Explorer 11 (should be useable, not pixel perfect)
- [x] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [ ] Firefox 68 (or latest version for Red Hat Enterprise Linux distribution)
- [x] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [x] Latest 2 versions of Safari
- [x] Android mobile device (such as the Galaxy S9)
- [x] Apple mobile device (such as the iPhone X)
- [x] Apple tablet device (such as the iPhone Pro)

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- [x] Tests have been updated to cover these changes.
- [x] Browser testing passed.
- [x] Repository compiles and tests pass.
- [x] Changelog updated.
- [x] Documentation (README.md, WHY.md, etc.) updated or added.
- [x] Link to the demo recording: []()
- [x] Approved by designer.

### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

### CODE REVIEW

File: elements/pfe-readtime/demo/index.html [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        <h1>&#x3C;pfe-readtime&#x3E;</h1>
+      </div>
+    </div>
+
+    <!-- Example 1 -->
+    <div class="container" data-padding="30">
+
+      <pfe-readtime for="readtime1" hidden>
+        <span slot="read-string">%t-minute readtime</span>
+        <span slot="read-string-less">Less than 1-minute readtime</span>
```
Comment (starryeyez024) [inline elements/pfe-readtime/demo/index.html:57]: Drive-by comment, we are probably not going to use this on RHDC. If something was so short that you could read it in less than a minute, probably it doesn't need the readtime component. It could just not render any text at all?

Comment (alazzara) [reply elements/pfe-readtime/demo/index.html:57]: sounds good. Thanks for the drive-by feedback.

File: elements/pfe-readtime/DISCOVERY.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
+# Project name
```
Comment (castastrophe) [inline elements/pfe-readtime/DISCOVERY.md:1]: You can delete this file if you did not need to use it for discovery

File: elements/pfe-readtime/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
+# PatternFly Element | Readtime element
+It takes in the word count of a given section and does a calculation on that number to return an estimated read time.
```
[the remark this answers is not in the document]
Comment (castastrophe) [reply elements/pfe-readtime/README.md:2]: I added a bit more information as this is what will be seen on the npm landing page for this component and will give users a bit more information to determine if this is the component they need.

File: elements/pfe-readtime/demo/pfe-readtime.story.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import { storiesOf } from "@storybook/polymer";
+import * as bridge from "@storybook/addon-knobs/polymer";
```
[the remark this answers is not in the document]
Comment (castastrophe) [reply elements/pfe-readtime/demo/pfe-readtime.story.js:2]: @alazzara This will fix your build error I believe ^

File: elements/pfe-readtime/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+## Usage
+
+```html
+    <!-- Default slot -->
+    <pfe-readtime for="readtime1">-minute readtime</pfe-readtime>
```
Comment (castastrophe) [inline elements/pfe-readtime/README.md:8]: I suspect this can be updated with examples utilizing the new slots we added: `read-string` and `read-string-less`

File: elements/pfe-readtime/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+### Accessibility
+Explain how this component meets accessibility standards.
```
Comment (castastrophe) [inline elements/pfe-readtime/README.md:12]: This is boilerplate text so it will need to be updated to include details about how this attribute meetings accessibility requirements.

File: elements/pfe-readtime/README.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+## Slots
+
+- `namedSlot`: There is only one slot and it contains a <span> with the readtime value.
```
Comment (castastrophe) [inline elements/pfe-readtime/README.md:16]: Slot details need to be updated

File: elements/pfe-readtime/README.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+## Attributes
+
+- `attr`: word-count is the data-attribute you will leverage for readtime. Example: word-count="2500" will let the component know that there is 2500 words in that section and it will do it's calculations based on that number. If you don't want to/have a data-attribute to leverage you can use id="readtime1" and it will get the word count for you.
```
Comment (castastrophe) [inline elements/pfe-readtime/README.md:20]: Need to update here to include the new attributes we worked on

File: elements/pfe-readtime/README.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+## Variable hooks
+
+Available hooks for styling:
+
+| Variable name | Default value | Region |
+| --- | --- | --- |
+| `--pfe-pfe-readtime--Color` | `#252527` | N/A |
```
Comment (castastrophe) [inline elements/pfe-readtime/README.md:28]: You can add:

```
:root { @include pfe-debug-log; }
```

in your sass file after the $LOCAL-VARIABLES map and copy the output in the _temp/pfe-readtime.css.  Be sure to delete that line before committing though.

File: elements/pfe-readtime/README.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+## Code style
+
+Readtime (and all PFElements) use [Prettier][prettier] to auto-format JS and JSON. The style rules get applied when you commit a change. If you choose to, you can [integrate your editor][prettier-ed] with Prettier to have the style rules applied on every save.
+
+[prettier]: https://github.com/prettier/prettier/
+[prettier-ed]: https://prettier.io/docs/en/editors.html
+[web-component-tester]: https://github.com/Polymer/web-component-tester
+
+## Readtime calculation information
```
Comment (castastrophe) [inline elements/pfe-readtime/README.md:63]: This is a great section, I'd bump this up above the dependencies section to make it more visible.

File: elements/pfe-readtime/demo/demo.css [hunk 11]
```diff
@@ -0,0 +1,31 @@
+body {
```
Comment (castastrophe) [inline elements/pfe-readtime/demo/demo.css:1]: You can leave this off, pfe-base should handle this for you

File: elements/pfe-readtime/demo/index.html [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+    <link href="../../pfelement/dist/pfelement.min.css" rel="stylesheet">
+
+    <!-- Stylesheets for testing light DOM styles.
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/css/bootstrap-reboot.css">
+    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/typebase.css/0.5.0/typebase.css">
+    -->
+
+    <link rel="stylesheet" href="../../pfe-styles/dist/pfe-base.css" />
+    <link rel="stylesheet" href="../../pfe-styles/dist/pfe-context.css" />
```
Comment (castastrophe) [inline elements/pfe-readtime/demo/index.html:20]: You can comment out pfe-context and pfe-layouts if you're not using them

File: elements/pfe-readtime/src/pfe-readtime.css [hunk 13]
```diff
@@ -0,0 +1 @@
+
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.css:1]: You can delete this file

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+      description:
+        "This element will collect a word count on a given section and calculate the readtime based on that count."
+    };
+  }
+
+  get templateUrl() {
+    return "pfe-readtime.html";
+  }
+
+  get styleUrl() {
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:20]: Since you don't have styles, I believe you can delete this whole section and it should be fine.

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+      }
+    };
+  }
+
+  static get slots() {
+    return {};
+  }
+
+  constructor() {
+    PFElement._debugLog = true;
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:65]: You can delete this now

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+    super(PfeReadtime, { type: PfeReadtime.PfeType, delayRender: true });
+
+    this.readStringTemplate = "%t-minute read";
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:68]: Maybe add a comment here describing this is the default value?

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+    }
+    this.log("_forChangedHandler");
+  }
+
+  _getWordCountOfSection(target) {
+    //WIP for API to leverage outside of data attribute in webrh
+    let wordCount;
+    if (target.hasAttribute("word-count")) {
+      wordCount = target.getAttribute("word-count");
+      console.log("Word count: " + wordCount);
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:103]: You can remove the console logs now or update then to `this.log`

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  //rename later
+  _readtime() {
+    if (this.readtime > 0) {
+      this.readString = this.readStringTemplate.replace("%t", this.readtime);
+    } else {
+      this.readString = "";
+    }
+    this.log("readtime");
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:187]: Can delete this now

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+
+    this.render();
+  }
+
+  //rename later
+  _readtime() {
+    if (this.readtime > 0) {
+      this.readString = this.readStringTemplate.replace("%t", this.readtime);
+    } else {
+      this.readString = "";
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:185]: Are we not rendering anything if the readtime is less than 1?

File: elements/pfe-readtime/test/index.html [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+  <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes">
+  <script src="/components/web-component-tester/browser.js"></script>
+</head>
+
+<body>
+  <script>
+    // Load and run all tests (.html, .js):
+    WCT.loadSuites([
+      'pfe-readtime_test.html',
+      // 'pfe-readtime_react_test.html',
```
Comment (castastrophe) [inline elements/pfe-readtime/test/index.html:15]: Need to comment this back in

File: elements/pfe-readtime/test/pfe-readtime_vue_test.html [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+  <script type="module" src="../dist/pfe-readtime.js"></script>
+</head>
+
+<body>
+  <div id="root"></div>
+  <script>
+    const app = new Vue({
+      el: "#root",
+      template: `
+    <pfe-readtime>
```
Comment (castastrophe) [inline elements/pfe-readtime/test/pfe-readtime_vue_test.html:19]: Need to update the vue and react markup to mimic the pfe-readtime_test.html content

File: elements/pfe-readtime/test/pfe-readtime_test.js [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+
+    test("it should upgrade", () => {
+        assert.instanceOf(
+            document.querySelector("pfe-readtime"),
+            customElements.get("pfe-readtime"),
+            "pfe-readtime should be an instance of pfeReadtime"
+        );
+    });
+
+    // Write tests for each attribute
```
Comment (castastrophe) [inline elements/pfe-readtime/test/pfe-readtime_test.js:13]: Can you start by stubbing out some comments for what tests need to be written and we can pair program on how to write these?

File: elements/pfe-readtime/src/pfe-readtime.js [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+
+  connectedCallback() {
+    super.connectedCallback();
+
+    this.slots = {
+      readString: this.querySelector(`[slot="read-string"]`)
+    };
+
+    //doesn't seem like this is needed
+    // if (this.slots.readString) this.readStringTemplate = this.slots.readString.textContent;
```
Comment (castastrophe) [inline elements/pfe-readtime/src/pfe-readtime.js:79]: The slots are no longer being used to generate the text for the result; only the default value is being picked up.  Can you let me know where you'd like to pull in the slotted content to use as a the template?

**Review by castastrophe — CHANGES_REQUESTED** [review]
Thanks for tackling this!

File: elements/pfe-readtime/test/pfe-readtime_test.js (hunk 22 above)
Comment (alazzara) [reply elements/pfe-readtime/test/pfe-readtime_test.js:13]: can do

File: elements/pfe-readtime/src/pfe-readtime.js (hunk 19 above)
Comment (alazzara) [reply elements/pfe-readtime/src/pfe-readtime.js:185]: correct. Based on Kendalls feedback.

File: elements/pfe-readtime/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alazzara) [reply elements/pfe-readtime/README.md:2]: cool. Thanks!

File: elements/pfe-readtime/src/pfe-readtime.js (hunk 19 above)
Comment (alazzara) [reply elements/pfe-readtime/src/pfe-readtime.js:185]: @castastrophe ^

File: elements/pfe-readtime/demo/index.html [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+      <div class="container" data-padding="30">
+        <h1>&#x3C;pfe-readtime&#x3E;</h1>
+      </div>
+    </div>
+
+    <!-- Example 1 -->
+    <div class="container" data-padding="30">
+
+      <pfe-readtime for="readtime1" id="pfeReadtime" hidden test>
+        <span slot="read-string">%t-minute readtime</span>
```
Comment (starryeyez024) [inline elements/pfe-readtime/demo/index.html:56]: Can we add a second demo with this:
```suggestion
        <span slot="read-string">Minutes to read: %t or so</span>
```
to demonstrate that it is `%t` which is being replaced and not the whole string?

File: elements/pfe-readtime/src/pfe-readtime.js (hunk 16 above)
Comment (starryeyez024) [reply elements/pfe-readtime/src/pfe-readtime.js:67]: I'm not sure if this code suggestion is exactly what is needed, but it should only be replacing the token `%t` with the number and not the whole string
```suggestion
    this.readStringTemplate = "%t";
```

File: elements/pfe-readtime/src/pfe-readtime.scss [hunk 25]
```diff
@@ -0,0 +1,8 @@
+// Please see the pfe-sass README for guidance on using these tools
+@import "../../pfe-sass/pfe-sass";
+
+$LOCAL: readtime;
+
+:host {
+  @include pfe-c-typography($type: text, $sizing: md);
```
Comment (starryeyez024) [inline elements/pfe-readtime/src/pfe-readtime.scss:7]: @alazzara  I saw this CSS which was good in the component, but I don't see it on the anbartle sandbox

**Review by castastrophe — APPROVED** [review]

**Review by castastrophe — APPROVED** [review]
